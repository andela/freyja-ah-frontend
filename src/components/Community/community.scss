$border: #333;
$body-color: #c4c4c4;

.main_com {
  display: flex;
  flex-direction: row;
  padding-left: 60px;

  .spinner-icon-container {
    height: 80px;
    left: 60%;
    position: absolute;
    top: 50%;
    transform: translate(-50%, -50%);
    width: 125px;
  }

  .spinner_overlay {
    position: fixed;
    width: 100%;
    height: 100%;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: rgba(0, 0, 0, 0.5);
    z-index: 2;
    cursor: pointer;
  }

  @media screen and (max-width: 700px) {
    padding-left: 0;
  }
  .d-aside {
    @media screen and (max-width: 500px) {
      display: none;
    }
    .c-side {
      padding-left: 45px;

      .c-logo {
        cursor: pointer;
        margin: 4rem 1rem;
        text-align: center;
        padding: 5px;
      }
    }
  }
  .main_sec {
    padding: 1rem;
    width: 90%;
    margin-top: 50px;

    .input_section {
      textarea,
      #inp {
        width: 60%;
        margin-top: 15px;
        border: 1px solid $border;

        &:focus {
          outline: none;
        }
      }

      .com_btn {
        background-color: $form-color;
        border: 0;
        border-radius: 5px;
        color: $white;
        width: 30%;
        margin-top: 10px;
        cursor: pointer;
        display: block;

        &:hover {
          opacity: 0.8;
          transition: 0.7s;
        }
      }

      hr {
        color: $border;
        border: 1px solid $border;
        margin-top: 40px;
      }
    }
    .posts {
      overflow: scroll;
      max-height: 40vh;

      .posts_section {
        padding: 1rem;

        .post_title {
          font-size: 20px;
          display: inline;
        }

        .timeCreated {
          display: inline-block;
          margin-left: 5%;
          color: $body-color;
        }

        .post_body {
          margin-top: 5px;
          color: $border;
        }

        .poster_name {
          display: inline-block;
          color: $form-color;
        }

        .poster_role {
          display: inline-block;
          padding-left: 1rem;
        }

        hr {
          border: 0.5px solid $form-color;
          margin: 0;
        }
      }
    }
  }
}
