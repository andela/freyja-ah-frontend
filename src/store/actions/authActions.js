import axios from 'axios';
import jwtDecode from 'jwt-decode';
import { GET_ERRORS, SET_CURRENT_USER, LOGIN_ERROR, INIT_AUTH_REQUEST, END_AUTH_REQUEST, VERIFY_USER } from './types';
import setAuthToken from '../../utils/setAuthToken';
import { baseUrl } from '../../utils/config';


/**
 * @method setCurrentUser
 * @description sets the current user of the app
 * @param {object} decoded user details
  * @returns {object}
 */
export const setCurrentUser = decoded => ({
  type: SET_CURRENT_USER,
  payload: decoded,
});

/**
 * @method setError
 * @description sets app error
 * @param {object} error
  * @returns {object}
 */
export const setError = error => ({
  type: GET_ERRORS,
  errors: error,
});

export const verifyUser = state => ({
  type: VERIFY_USER,
  isVerified: state,
});

/**
 * @method registerUser
 * @description registers a new user
 * @param {object} newUser
  * @returns {object} user object
 */
export const registerUser = newUser => async (dispatch) => {
  try {
    const user = await axios.post(`${baseUrl}/users`, newUser);

    if (user) {
      const { token } = user.data;
      localStorage.setItem('token', token);
      const decoded = jwtDecode(token);
      dispatch(setCurrentUser(decoded));
    }
  } catch (err) {
    const { error } = err.response.data;
    dispatch(setError(error));
  }
};

/**
 * @method loginUser
 * @description registers a new user
 * @param {object} payload
  * @returns {object} user object
 */
export const loginUser = payload => async (dispatch) => {
  dispatch({ type: INIT_AUTH_REQUEST });

  try {
    const request = await axios.post(`${baseUrl}/users/login`, payload);
    dispatch({
      type: END_AUTH_REQUEST,
    });

    const { token } = request.data;
    localStorage.setItem('token', token);

    setAuthToken(token);
    const decoded = jwtDecode(token);
    dispatch(setCurrentUser(decoded));
  } catch (error) {
    dispatch({
      type: END_AUTH_REQUEST,
    });

    const { data } = error.response;
    const errorObj = Array.isArray(data.error) ? { error: data.error[0].msg } : data;
    dispatch({
      type: LOGIN_ERROR,
      payload: errorObj,
    });
  }
};


/**
 * @method verifyAuthUser
 * @description registers a new user
 * @param {string} token
  * @returns {string} confirmation message
 */
export const verifyAuthUser = token => async (dispatch) => {
  try {
    await axios.get(`${baseUrl}/user/verify/${token}`);
    await dispatch(verifyUser("you've been verified, Redirecting to your profile page..."));
  } catch (error) {
    dispatch(verifyUser("Couldn't verify you, Please chcek the link sent to your email again."));
  }
};
