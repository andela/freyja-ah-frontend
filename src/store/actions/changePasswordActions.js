import axios from 'axios';
import { baseUrl } from '../../utils/config';
import { PASSWORD_CHANGE_ERROR, PASSWORD_CHANGE_SUCCESS, LOADING } from './types';

/**
 * @method loading
 * @description initiates loading
  * @returns {object}
 */
export const loading = () => ({
  type: LOADING,
});

/**
 * @method passwordChangeSuccess
 * @description handles password change success
 * @param {string} data success message
  * @returns {object}
 */
export const passwordChangeSuccess = data => ({
  type: PASSWORD_CHANGE_SUCCESS,
  payload: data,
});

/**
 * @method passwordChangeError
 * @description handles password change error
 * @param {string} data error message
  * @returns {object}
 */
export const passwordChangeError = data => ({
  type: PASSWORD_CHANGE_ERROR,
  payload: data,
});

/**
 * @method ChangePassword
 * @description change user's password
 * @param {string} password
 * @param {string} token
  * @returns {object}
 */
export const ChangePassword = (password, token) => async (dispatch) => {
  try {
    dispatch(loading());

    const resetPassword = await axios.post(
      `${baseUrl}/users/change-password?token=${token}`,
      { newPassword: password },
    );

    dispatch(passwordChangeSuccess(resetPassword.data.message));
  } catch (error) {
    dispatch(passwordChangeError('please send reset password request again'));
  }
};
