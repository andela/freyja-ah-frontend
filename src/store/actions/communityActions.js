import { get } from 'axios';
import { GET_COMMUNITY_MESSAGES, GET_COMMUNITY_MESSAGES_ERROR, LOADING } from './types';
import { baseUrl } from '../../utils/config';

/**
 * @method communityMessageSuccess
 * @param {string} data success message
  * @returns {object}
 */
export const communityMessageSuccess = data => ({
  type: GET_COMMUNITY_MESSAGES,
  data,
});

/**
 * @method communityMessageSuccess
 * @param {string} data error message
  * @returns {object}
 */
export const communityMessageError = data => ({
  type: GET_COMMUNITY_MESSAGES_ERROR,
  data,
});

/**
 * @method loading
 * @description initiates loading
  * @returns {object}
 */
export const loading = () => ({
  type: LOADING,
});

/**
 * @method communityMessageSuccess
 * @description get all community messages
  * @returns {object}
 */
export const getCommunityMessages = () => async (dispatch) => {
  dispatch(loading());
  try {
    const response = await get(`${baseUrl}/community/messages`, {
      headers: { Authorization: `${localStorage.getItem('token')}` },
    });
    dispatch(communityMessageSuccess(response.data.data));
  } catch (error) {
    dispatch(communityMessageError('internal error'));
  }
};
