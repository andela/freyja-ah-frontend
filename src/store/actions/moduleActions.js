import { get } from 'axios';
import {
  GET_ALL_MODULES, GET_SINGLE_MODULE, MODULES_ERROR, INIT_MODULE_REQUEST,
} from './types';
import { baseUrl } from '../../utils/config';

/**
 * @method getModules
 * @description get all course modules
  * @returns {object}
 */
export const getModules = () => async (dispatch) => {
  try {
    const modulesRequest = await get(`${baseUrl}/modules`, {
      headers: { Authorization: `${localStorage.getItem('token')}` },
    });

    const { data } = modulesRequest.data;
    return dispatch({
      type: GET_ALL_MODULES,
      payload: data,
    });
  } catch (error) {
    return dispatch({
      type: MODULES_ERROR,
      payload: error,
    });
  }
};

/**
 * @method setModules
 * @description set current module
 * @param {object} payload module object
  * @returns {object}
 */
export const setModule = payload => ({
  type: GET_SINGLE_MODULE,
  payload,
});

export const getModule = moduleId => async (dispatch) => {
  try {
    dispatch({ type: INIT_MODULE_REQUEST });
    const request = await get(`${baseUrl}/modules/${moduleId}`);
    const { data } = request.data;
    dispatch(setModule(data));
  } catch (error) {
    dispatch({
      type: MODULES_ERROR,
      payload: error,
    });
  }
};
