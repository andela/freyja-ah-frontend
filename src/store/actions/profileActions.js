import axios from 'axios';
import * as actionCreators from './types';
import { baseUrl } from '../../utils/config';

const token = localStorage.getItem('token');

/**
 * @method getProfileSuccess
 * @param {object} profile profile object
  * @returns {object}
 */
export const getProfileSuccess = profile => ({
  type: actionCreators.GET_PROFILE_SUCCESS,
  data: profile,
});

/**
 * @method getProfileFailed
 * @param {string} error profile object
  * @returns {object}
 */
export const getProfileFailed = error => ({
  type: actionCreators.GET_PROFILE_FAILED,
  error,
});

/**
 * @method  getProfile
 * @param {string} userId user id
  * @returns {object}
 */
export const getProfile = userId => async (dispatch) => {
  const config = {
    headers: {
      Authorization: token,
    },
  };
  try {
    const profile = await axios.get(`${baseUrl}/profiles/${userId}`, config);
    const { data } = profile;
    const { user } = data;
    let dob = null;
    if (user.profile.dateOfBirth) {
      const dobArray = (user.profile.dateOfBirth).split('T');
      [dob] = dobArray;
    }

    const userDetails = {
      name: `${user.firstName} ${user.lastName}`,
      email: user.email,
      bio: user.profile.bio,
      image: user.profile.image,
      username: user.userName,
      phoneNumber: user.profile.phoneNumber,
      socialMedia: [
        { name: 'facebook', url: user.profile.facebook },
        { name: 'twitter', url: user.profile.twitter },
        { name: 'linkedin', url: user.profile.linkedIn },
      ],
      DOB: dob,
      industry: user.profile.industry,
      yrsOfExperience: user.profile.yrsOfExperience,
      isEnrolled: user.profile.isEnrolled,
    };

    dispatch(getProfileSuccess(userDetails));
  } catch (error) {
    dispatch(getProfileFailed(error));
  }
};

/**
 * @method  uploadStart
 * @description initiates image upload
  * @returns {object}
 */
export const uploadStart = () => ({
  type: actionCreators.UPLOAD_START,
});

/**
 * @method  uploadSuccess
 * @description handles image upload success
 * @param {string} imageUrl upload image url
  * @returns {object}
 */
export const uploadSuccess = imageUrl => ({
  type: actionCreators.UPLOAD_SUCCESS,
  imageUrl,
});

/**
 * @method  uploadFailed
 * @description handles image upload failure
 * @param {object} error error object
  * @returns {object}
 */
export const uploadFailed = error => ({
  type: actionCreators.UPLOAD_FAILED,
  error,
});

/**
 * @method  uploadImage
 * @description handles image upload
 * @param {file} image image
  * @returns {object}
 */
export const uploadImage = image => async (dispatch) => {
  dispatch(uploadStart());
  const config = {
    headers: {
      'Content-type': 'multipart/form-data',
      Authorization: token,
    },
  };
  try {
    const data = new FormData();
    data.append('image', image);

    const upload = await axios.post(`${baseUrl}/image`, data, config);
    const imageUrl = upload.data.data;
    dispatch(uploadSuccess(imageUrl));
  } catch (error) {
    dispatch(uploadFailed(error.response.data));
  }
};

/**
 * @method  updateProfileFailed
 * @description handles profile update failure
 * @param {object} error update profile error
  * @returns {object}
 */
export const updateProfileFailed = error => ({
  type: actionCreators.UPDATE_PROFILE_FAILED,
  error,
});

/**
 * @method  updateProfile
 * @description handles profile update
 * @param {object} data profile data
  * @returns {object}
 */
export const updateProfile = data => async (dispatch) => {
  const config = {
    headers: {
      Authorization: token,
    },
  };
  try {
    const profile = await axios.put(`${baseUrl}/profiles`, data, config);
    await dispatch(getProfile(profile.data.profile.userId));
  } catch (error) {
    dispatch(updateProfileFailed(error.response.data.error));
  }
};
