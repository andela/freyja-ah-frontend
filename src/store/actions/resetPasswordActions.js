import axios from 'axios';
import { baseUrl } from '../../utils/config';
import { PASSWORD_RESET_SUCCESS, PASSWORD_RESET_ERROR, LOADING } from './types';

/**
 * @method loading
 * @description initiates loading
  * @returns {object}
 */
export const loading = () => ({
  type: LOADING,
});

/**
 * @method passwordResetSuccess
 * @description handles password reset success
 * @param {string} data success message
  * @returns {object}
 */
export const passwordResetSuccess = data => ({
  type: PASSWORD_RESET_SUCCESS,
  payload: data,
});

/**
 * @method passwordResetError
 * @description handles password reset error
 * @param {string} data error message
  * @returns {object}
 */
export const passwordResetError = data => ({
  type: PASSWORD_RESET_ERROR,
  payload: data,
});

/**
 * @method ResetPassword
 * @description reset user's password
 * @param {string} email
  * @returns {object}
 */
export const ResetPassword = email => async (dispatch) => {
  try {
    dispatch(loading());

    const resetPassword = await axios.post(`${baseUrl}/users/reset`, { email });

    dispatch(passwordResetSuccess(resetPassword.data.message));
  } catch (error) {
    dispatch(passwordResetError(error.response.data.message));
  }
};
