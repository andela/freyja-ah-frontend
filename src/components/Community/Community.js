import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { withRouter, Link } from 'react-router-dom';
import { connect } from 'react-redux';
import Moment from 'react-moment';
import Header from '../Header/Header';
import Spinner from '../Spinner/Spinner';
import Footer from '../Footer/Footer';
import Sidebar from '../Sidebar/Sidebar';
import Input from '../Inputs/Input';
import Button from '../Button';
import * as actions from '../../store/actions/communityActions';
import './community.scss';
import sideLogo from '../../assets/images/logo.png';

export class Community extends PureComponent {
  state = {
    messages: [],
  };

  componentDidMount() {
    const getCommMessages = async () => {
      const { getCommunityMessages } = this.props;
      await getCommunityMessages();
      const { messages } = this.props;
      this.setState({ messages });
    };
    getCommMessages();
  }

  render() {
    const spinner = (
      <div className="spinner-position">
        <Spinner />
      </div>
    );

    const { messages } = this.state;
    const { loading } = this.props;
    return (
      <div>
        <Header />
        <div className="main_com">
          <Sidebar>
            <div className="d-aside">
              <div className="c-side">
                <span className="c-logo">
                  <img src={sideLogo} alt="logo" />
                  Community
                </span>
              </div>
              <ul className="d-links">
                <li className="module-active">
                  <Link to="/community">Forum</Link>
                </li>
                <li>
                  <Link to="/profile">Profile</Link>
                </li>
                <li>
                  <Link to="/payment">Messages</Link>
                </li>
                <li>
                  <Link to="/aboutus">Help</Link>
                </li>
              </ul>
            </div>
          </Sidebar>
          <div className="main_sec">
            <div className="input_section">
              <Input id="inp" placeholder="Title" />
              <textarea rows="3" placeholder="enter your message..." />
              <Button text="Post message" classname="com_btn" />
              <hr />
            </div>
            <div className="posts">
              {loading ? (
                spinner
              ) : (
                messages.map(msg => (
                  <div className="posts_section" key={msg && msg.id}>
                    <h4 className="post_title">
                      {msg.owner && msg.owner.lastName}
                      {' '}
                      {msg.owner && msg.owner.firstName}
                    </h4>
                    <p className="timeCreated">
                      <Moment local format="LLLL">
                        {msg.createdAt}
                      </Moment>
                    </p>
                    <p className="post_body">{msg.body}</p>
                    <p className="poster_name">{msg.owner && msg.owner.userName}</p>
                    <p className="poster_role">{msg.owner && msg.owner.role}</p>
                    <hr />
                  </div>
                ))
              )}
            </div>
          </div>
        </div>
        <Footer />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  loading: state.community.isLoading,
  messages: state.community.allMessages,
  error: state.community.error,
});

const mapDispatchToProps = {
  getCommunityMessages: () => actions.getCommunityMessages(),
};

Community.propTypes = {
  getCommunityMessages: PropTypes.func,
  loading: PropTypes.bool,
  messages: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    body: PropTypes.string,
  })),
};

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(Community));
