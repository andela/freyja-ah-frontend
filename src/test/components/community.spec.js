import React from 'react';
import { shallow } from 'enzyme';
import { Community } from '../../components/Community/Community';
import * as actions from '../../store/actions/communityActions';

describe('test community component methods', () => {
  const props = {
    getCommunityMessages: () => actions.getCommunityMessages(),
  };

  const enzymeWrapper = shallow(<Community {...props} />);
  let wrapper;
  beforeEach(() => {
    wrapper = enzymeWrapper;
  });

  it('should test community', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
