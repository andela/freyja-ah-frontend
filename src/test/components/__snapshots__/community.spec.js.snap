// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test community component methods should test community 1`] = `ShallowWrapper {}`;
